INTERNAL ERROR (@classmethod shadows @dataclass)

Given `test.py` as: 
```
from dataclasses import dataclass

@dataclass
class Foo:
    bar: str

    @classmethod
    def bar(cls) -> "Foo":
        return cls()
```

and 
```
> python --version
Python 3.7.0
```

and 

```
> pip freeze | grep mypy
mypy==0.660
mypy-extensions==0.4.1
```

I get:

```
> mypy test.py --show-traceback
test.py:4: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues version: 0.660
Traceback (most recent call last):
  File "/Users/jessemyers/.venvs/microcosm-neo4j/bin/mypy", line 11, in <module>
    sys.exit(console_entry())
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/__main__.py", line 7, in console_entry
    main(None)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/main.py", line 91, in main
    res = build.build(sources, options, None, flush_errors, fscache)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/build.py", line 157, in build
    result = _build(sources, options, alt_lib_path, flush_errors, fscache)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/build.py", line 212, in _build
    graph = dispatch(sources, manager)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/build.py", line 2246, in dispatch
    process_graph(graph, manager)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/build.py", line 2544, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/build.py", line 2653, in process_stale_scc
    graph[id].semantic_analysis()
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/build.py", line 1780, in semantic_analysis
    self.manager.semantic_analyzer.visit_file(self.tree, self.xpath, self.options, patches)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/semanal.py", line 303, in visit_file
    self.accept(d)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/semanal.py", line 3708, in accept
    node.accept(self)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/nodes.py", line 851, in accept
    return visitor.visit_class_def(self)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/semanal.py", line 800, in visit_class_def
    defn.defs.accept(self)
  File "/Users/jessemyers/.pyenv/versions/3.7.0/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/semanal.py", line 854, in analyze_class_body
    self.apply_class_plugin_hooks(defn)
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/semanal.py", line 880, in apply_class_plugin_hooks
    hook(ClassDefContext(defn, decorator, self))
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/plugins/dataclasses.py", line 312, in dataclass_class_maker_callback
    transformer.transform()
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/plugins/dataclasses.py", line 79, in transform
    attributes = self.collect_attributes()
  File "/Users/jessemyers/.venvs/microcosm-neo4j/lib/python3.7/site-packages/mypy/plugins/dataclasses.py", line 189, in collect_attributes
    assert isinstance(node, Var)
AssertionError:
test.py:4: : note: use --pdb to drop into pdb
```

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -495,6 +495,19 @@ reveal_type(A.foo("foo"))  # N: Revealed type is "builtins.str"
 
 [builtins fixtures/dataclasses.pyi]
 
+[case testClassmethodShadowingFieldDoesNotCrash]
+# flags: --python-version 3.7
+from dataclasses import dataclass
+
+# This used to crash -- see #6217
+@dataclass
+class Foo:
+    bar: str
+    @classmethod  # E: Name "bar" already defined on line 7
+    def bar(cls) -> "Foo":
+        return cls('asdf')
+[builtins fixtures/dataclasses.pyi]
+
 [case testDataclassesClassVars]
 # flags: --python-version 3.7
 from dataclasses import dataclass

